src/compiler/checker.ts

```diff
@@ -6703,8 +6790,11 @@ module ts {
                 }
                 else {
                     var exprType = checkExpression(<Expression>node.body);
-                    if (node.type) {
-                        checkTypeAssignableTo(exprType, getTypeFromTypeNode(node.type), node.body, /*headMessage*/ undefined);
+                    if (returnType) {
+                        if (isAsync) {
+                            exprType = getAwaitedType(exprType, /*fallbackType*/ exprType);
+                    }
+                        checkTypeAssignableTo(exprType, returnType, node.body, /*headMessage*/ undefined);

```

Comment:
But in that case, why would you compare to returnType? Don't you need to compare to the promised type of returnType?